File: examples/jog_axis.py
#!/usr/bin/env python3
"""
Jog Axis Example

Demonstrates jogging an axis using the LinuxCNC gRPC server.
Supports both continuous jogging and incremental jogging.

Usage:
    python jog_axis.py [--host HOST] [--port PORT]

Safety Warning:
    This script moves the machine! Ensure you have clear access to E-stop
    and understand the jog parameters before running.
"""

import argparse
import sys
import time

import grpc

# Try installed package first, fall back to local src/ directory
try:
    from linuxcnc_grpc_server._generated import linuxcnc_pb2
    from linuxcnc_grpc_server._generated import linuxcnc_pb2_grpc
except ImportError:
    from pathlib import Path
    sys.path.insert(0, str(Path(__file__).parent.parent / "src"))
    from linuxcnc_grpc_server._generated import linuxcnc_pb2
    from linuxcnc_grpc_server._generated import linuxcnc_pb2_grpc


class LinuxCNCClient:
    """Simple client wrapper for LinuxCNC gRPC operations."""

    def __init__(self, host: str, port: int):
        self.channel = grpc.insecure_channel(f"{host}:{port}")
        self.stub = linuxcnc_pb2_grpc.LinuxCNCServiceStub(self.channel)
        self._serial = 0

    def close(self):
        self.channel.close()

    def _next_serial(self) -> int:
        self._serial += 1
        return self._serial

    def get_status(self) -> linuxcnc_pb2.LinuxCNCStatus:
        """Get current machine status."""
        return self.stub.GetStatus(linuxcnc_pb2.GetStatusRequest())

    def send_command(self, command: linuxcnc_pb2.LinuxCNCCommand) -> linuxcnc_pb2.CommandResponse:
        """Send a command and return the response."""
        command.serial = self._next_serial()
        command.timestamp = int(time.time() * 1e9)
        return self.stub.SendCommand(command)

    def set_mode(self, mode: linuxcnc_pb2.TaskMode) -> linuxcnc_pb2.CommandResponse:
        """Set task mode (MANUAL, AUTO, MDI)."""
        cmd = linuxcnc_pb2.LinuxCNCCommand()
        cmd.mode.mode = mode
        return self.send_command(cmd)

    def set_state(self, state: linuxcnc_pb2.TaskState) -> linuxcnc_pb2.CommandResponse:
        """Set machine state (ESTOP, ESTOP_RESET, ON, OFF)."""
        cmd = linuxcnc_pb2.LinuxCNCCommand()
        cmd.state.state = state
        return self.send_command(cmd)

    def jog_continuous(self, axis: int, velocity: float) -> linuxcnc_pb2.CommandResponse:
        """Start continuous jog on an axis. velocity > 0 for positive, < 0 for negative."""
        cmd = linuxcnc_pb2.LinuxCNCCommand()
        cmd.jog.type = linuxcnc_pb2.JOG_CONTINUOUS
        cmd.jog.is_joint = False  # Jog in axis (teleop) mode
        cmd.jog.index = axis      # 0=X, 1=Y, 2=Z, etc.
        cmd.jog.velocity = velocity
        return self.send_command(cmd)

    def jog_increment(self, axis: int, velocity: float, increment: float) -> linuxcnc_pb2.CommandResponse:
        """Jog axis by a fixed increment."""
        cmd = linuxcnc_pb2.LinuxCNCCommand()
        cmd.jog.type = linuxcnc_pb2.JOG_INCREMENT
        cmd.jog.is_joint = False
        cmd.jog.index = axis
        cmd.jog.velocity = abs(velocity)
        cmd.jog.increment = increment  # Signed increment
        return self.send_command(cmd)

    def jog_stop(self, axis: int) -> linuxcnc_pb2.CommandResponse:
        """Stop jogging on an axis."""
        cmd = linuxcnc_pb2.LinuxCNCCommand()
        cmd.jog.type = linuxcnc_pb2.JOG_STOP
        cmd.jog.is_joint = False
        cmd.jog.index = axis
        cmd.jog.velocity = 0
        return self.send_command(cmd)


def ensure_machine_ready(client: LinuxCNCClient) -> bool:
    """Ensure machine is out of E-stop, powered on, and in manual mode."""
    status = client.get_status()

    # Check E-stop
    if status.task.task_state == linuxcnc_pb2.STATE_ESTOP:
        print("Machine is in E-stop. Resetting...")
        resp = client.set_state(linuxcnc_pb2.STATE_ESTOP_RESET)
        if resp.status != linuxcnc_pb2.RCS_DONE:
            print(f"Failed to reset E-stop: {resp.error_message}")
            return False
        time.sleep(0.1)

    # Power on
    status = client.get_status()
    if status.task.task_state != linuxcnc_pb2.STATE_ON:
        print("Powering on machine...")
        resp = client.set_state(linuxcnc_pb2.STATE_ON)
        if resp.status != linuxcnc_pb2.RCS_DONE:
            print(f"Failed to power on: {resp.error_message}")
            return False
        time.sleep(0.1)

    # Set manual mode for jogging
    status = client.get_status()
    if status.task.task_mode != linuxcnc_pb2.MODE_MANUAL:
        print("Setting manual mode...")
        resp = client.set_mode(linuxcnc_pb2.MODE_MANUAL)
        if resp.status != linuxcnc_pb2.RCS_DONE:
            print(f"Failed to set manual mode: {resp.error_message}")
            return False
        time.sleep(0.1)

    return True


def demo_incremental_jog(client: LinuxCNCClient):
    """Demonstrate incremental jogging."""
    print("\n--- Incremental Jog Demo ---")
    print("Jogging X axis +1.0 units...")

    # Jog X axis positive by 1.0 unit at 100 units/min
    resp = client.jog_increment(axis=0, velocity=100.0, increment=1.0)
    if resp.status != linuxcnc_pb2.RCS_DONE:
        print(f"Jog failed: {resp.error_message}")
        return

    # Wait for motion to complete
    time.sleep(1.0)

    # Show new position
    status = client.get_status()
    pos = status.position.actual_position
    print(f"New position: X={pos.x:.4f} Y={pos.y:.4f} Z={pos.z:.4f}")


def demo_continuous_jog(client: LinuxCNCClient):
    """Demonstrate continuous jogging with stop."""
    print("\n--- Continuous Jog Demo ---")
    print("Jogging Y axis positive for 0.5 seconds...")

    # Start continuous jog on Y axis at 50 units/min
    resp = client.jog_continuous(axis=1, velocity=50.0)
    if resp.status != linuxcnc_pb2.RCS_DONE:
        print(f"Jog start failed: {resp.error_message}")
        return

    # Let it jog for a bit
    time.sleep(0.5)

    # Stop the jog
    print("Stopping jog...")
    resp = client.jog_stop(axis=1)
    if resp.status != linuxcnc_pb2.RCS_DONE:
        print(f"Jog stop failed: {resp.error_message}")
        return

    # Show new position
    status = client.get_status()
    pos = status.position.actual_position
    print(f"New position: X={pos.x:.4f} Y={pos.y:.4f} Z={pos.z:.4f}")


def main():
    parser = argparse.ArgumentParser(description="Jog axis demo via gRPC")
    parser.add_argument("--host", default="localhost", help="gRPC server host")
    parser.add_argument("--port", type=int, default=50051, help="gRPC server port")
    parser.add_argument("--skip-demo", action="store_true",
                        help="Skip the demo movements (just show status)")
    args = parser.parse_args()

    client = LinuxCNCClient(args.host, args.port)

    try:
        # Show initial status
        status = client.get_status()
        pos = status.position.actual_position
        print(f"Current position: X={pos.x:.4f} Y={pos.y:.4f} Z={pos.z:.4f}")

        if args.skip_demo:
            print("Skipping demo movements (--skip-demo)")
            return

        # Ensure machine is ready for jogging
        if not ensure_machine_ready(client):
            print("Could not prepare machine for jogging")
            sys.exit(1)

        # Run demos
        demo_incremental_jog(client)
        demo_continuous_jog(client)

        print("\nJog demo complete!")

    except grpc.RpcError as e:
        print(f"gRPC error: {e.code()}: {e.details()}", file=sys.stderr)
        sys.exit(1)
    finally:
        client.close()


if __name__ == "__main__":
    main()

File: examples/node/jog_axis.ts
/**
 * Jog Axis Example
 *
 * Demonstrates jogging an axis using the LinuxCNC gRPC server.
 * Supports both continuous jogging and incremental jogging.
 *
 * Usage:
 *   npx tsx jog_axis.ts [--host HOST] [--port PORT] [--skip-demo]
 *
 * Safety Warning:
 *   This script moves the machine! Ensure you have clear access to E-stop
 *   and understand the jog parameters before running.
 */

import * as grpc from "@grpc/grpc-js";
import { program } from "commander";
import {
  LinuxCNCServiceClient,
  GetStatusRequest,
  LinuxCNCCommand,
  LinuxCNCStatus,
  CommandResponse,
  TaskMode,
  TaskState,
  JogType,
  RcsStatus,
} from "linuxcnc-grpc";

program
  .option("--host <host>", "gRPC server host", "localhost")
  .option("--port <port>", "gRPC server port", "50051")
  .option("--skip-demo", "Skip demo movements (just show status)")
  .parse();

const opts = program.opts();
const address = `${opts.host}:${opts.port}`;

const client = new LinuxCNCServiceClient(
  address,
  grpc.credentials.createInsecure()
);

let serial = 0;

function nextSerial(): number {
  return ++serial;
}

function getStatus(): Promise<LinuxCNCStatus> {
  return new Promise((resolve, reject) => {
    client.getStatus(GetStatusRequest.create(), (err, status) => {
      if (err) reject(err);
      else resolve(status);
    });
  });
}

function sendCommand(cmd: LinuxCNCCommand): Promise<CommandResponse> {
  cmd.serial = nextSerial();
  cmd.timestamp = Date.now() * 1000000;
  return new Promise((resolve, reject) => {
    client.sendCommand(cmd, (err, response) => {
      if (err) reject(err);
      else resolve(response);
    });
  });
}

function setMode(mode: TaskMode): Promise<CommandResponse> {
  const cmd = LinuxCNCCommand.create();
  cmd.mode = { mode };
  return sendCommand(cmd);
}

function setState(state: TaskState): Promise<CommandResponse> {
  const cmd = LinuxCNCCommand.create();
  cmd.state = { state };
  return sendCommand(cmd);
}

function jogContinuous(axis: number, velocity: number): Promise<CommandResponse> {
  const cmd = LinuxCNCCommand.create();
  cmd.jog = {
    type: JogType.JOG_CONTINUOUS,
    isJoint: false,
    index: axis,
    velocity,
    increment: 0,
  };
  return sendCommand(cmd);
}

function jogIncrement(
  axis: number,
  velocity: number,
  increment: number
): Promise<CommandResponse> {
  const cmd = LinuxCNCCommand.create();
  cmd.jog = {
    type: JogType.JOG_INCREMENT,
    isJoint: false,
    index: axis,
    velocity: Math.abs(velocity),
    increment,
  };
  return sendCommand(cmd);
}

function jogStop(axis: number): Promise<CommandResponse> {
  const cmd = LinuxCNCCommand.create();
  cmd.jog = {
    type: JogType.JOG_STOP,
    isJoint: false,
    index: axis,
    velocity: 0,
    increment: 0,
  };
  return sendCommand(cmd);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function ensureMachineReady(): Promise<boolean> {
  let status = await getStatus();

  // Check E-stop
  if (status.task!.taskState === TaskState.STATE_ESTOP) {
    console.log("Machine is in E-stop. Resetting...");
    const resp = await setState(TaskState.STATE_ESTOP_RESET);
    if (resp.status !== RcsStatus.RCS_DONE) {
      console.log(`Failed to reset E-stop: ${resp.errorMessage}`);
      return false;
    }
    await sleep(100);
  }

  // Power on
  status = await getStatus();
  if (status.task!.taskState !== TaskState.STATE_ON) {
    console.log("Powering on machine...");
    const resp = await setState(TaskState.STATE_ON);
    if (resp.status !== RcsStatus.RCS_DONE) {
      console.log(`Failed to power on: ${resp.errorMessage}`);
      return false;
    }
    await sleep(100);
  }

  // Set manual mode for jogging
  status = await getStatus();
  if (status.task!.taskMode !== TaskMode.MODE_MANUAL) {
    console.log("Setting manual mode...");
    const resp = await setMode(TaskMode.MODE_MANUAL);
    if (resp.status !== RcsStatus.RCS_DONE) {
      console.log(`Failed to set manual mode: ${resp.errorMessage}`);
      return false;
    }
    await sleep(100);
  }

  return true;
}

async function demoIncrementalJog(): Promise<void> {
  console.log("\n--- Incremental Jog Demo ---");
  console.log("Jogging X axis +1.0 units...");

  // Jog X axis positive by 1.0 unit at 100 units/min
  const resp = await jogIncrement(0, 100.0, 1.0);
  if (resp.status !== RcsStatus.RCS_DONE) {
    console.log(`Jog failed: ${resp.errorMessage}`);
    return;
  }

  // Wait for motion to complete
  await sleep(1000);

  // Show new position
  const status = await getStatus();
  const pos = status.position!.actualPosition!;
  console.log(`New position: X=${pos.x.toFixed(4)} Y=${pos.y.toFixed(4)} Z=${pos.z.toFixed(4)}`);
}

async function demoContinuousJog(): Promise<void> {
  console.log("\n--- Continuous Jog Demo ---");
  console.log("Jogging Y axis positive for 0.5 seconds...");

  // Start continuous jog on Y axis at 50 units/min
  let resp = await jogContinuous(1, 50.0);
  if (resp.status !== RcsStatus.RCS_DONE) {
    console.log(`Jog start failed: ${resp.errorMessage}`);
    return;
  }

  // Let it jog for a bit
  await sleep(500);

  // Stop the jog
  console.log("Stopping jog...");
  resp = await jogStop(1);
  if (resp.status !== RcsStatus.RCS_DONE) {
    console.log(`Jog stop failed: ${resp.errorMessage}`);
    return;
  }

  // Show new position
  const status = await getStatus();
  const pos = status.position!.actualPosition!;
  console.log(`New position: X=${pos.x.toFixed(4)} Y=${pos.y.toFixed(4)} Z=${pos.z.toFixed(4)}`);
}

async function main(): Promise<void> {
  try {
    // Show initial status
    const status = await getStatus();
    const pos = status.position!.actualPosition!;
    console.log(`Current position: X=${pos.x.toFixed(4)} Y=${pos.y.toFixed(4)} Z=${pos.z.toFixed(4)}`);

    if (opts.skipDemo) {
      console.log("Skipping demo movements (--skip-demo)");
      return;
    }

    // Ensure machine is ready for jogging
    if (!(await ensureMachineReady())) {
      console.log("Could not prepare machine for jogging");
      process.exit(1);
    }

    // Run demos
    await demoIncrementalJog();
    await demoContinuousJog();

    console.log("\nJog demo complete!");
  } catch (err) {
    console.error(`Error: ${err}`);
    process.exit(1);
  } finally {
    client.close();
  }
}

main();
